src/java.base/share/classes/java/lang/foreign/MemoryInspection.java

```diff
@@ -0,0 +1,210 @@
+        }
+
+        /**
+         * Renders the provided {@code charLayout} and {@code value} to a String.
+         *
+         * @param charLayout the layout to render
+         * @param value      the value to render
+         * @return rendered String
+         */
+        default String render(ValueLayout.OfChar charLayout, char value) {
+            requireNonNull(charLayout);
+            return Character.toString(value);
+        }
+
+        /**
+         * Renders the provided {@code sortLayout} and {@code value} to a String.
+         *
+         * @param sortLayout the layout to render
+         * @param value      the value to render
+         * @return rendered String
+         */
+        default String render(ValueLayout.OfShort sortLayout, short value) {
+            requireNonNull(sortLayout);
+            return Short.toString(value);
+        }
+
+        /**
+         * Renders the provided {@code intLayout} and {@code value} to a String.
+         *
+         * @param intLayout the layout to render
+         * @param value     the value to render
+         * @return rendered String
+         */
+        default String render(ValueLayout.OfInt intLayout, int value) {
+            requireNonNull(intLayout);
+            return Integer.toString(value);
+        }
+
+        /**
+         * Renders the provided {@code longLayout} and {@code value} to a String.
+         *
+         * @param longLayout the layout to render
+         * @param value      the value to render
+         * @return rendered String
+         */
+        default String render(ValueLayout.OfLong longLayout, long value) {
+            requireNonNull(longLayout);
+            return Long.toString(value);
+        }
+
+        /**
+         * Renders the provided {@code floatLayout} and {@code value} to a String.
+         *
+         * @param floatLayout the layout to render
+         * @param value       the value to render
+         * @return rendered String
+         */
+        default String render(ValueLayout.OfFloat floatLayout, float value) {
+            requireNonNull(floatLayout);
+            return Float.toString(value);
+        }
+
+        /**
+         * Renders the provided {@code doubleLayout} and {@code value} to a String.
+         *
+         * @param doubleLayout the layout to render
+         * @param value        the value to render
+         * @return rendered String
+         */
+        default String render(ValueLayout.OfDouble doubleLayout, double value) {
+            requireNonNull(doubleLayout);
+            return Double.toString(value);
+        }
+
+        /**
+         * Renders the provided {@code addressLayout} and {@code value} to a String.
+         *
+         * @param addressLayout the layout to render
+         * @param value         the value to render
+         * @return rendered String
+         */
+        default String render(ValueLayout.OfAddress addressLayout, MemorySegment value) {
```

Comment:
I think you can make it take `Addressable` so it works for both `MemorySegment` and `MemoryAddress`:

```java
default String render(ValueLayout.OfAddress addressLayout, Addressable value) {
    requireNonNull(addressLayout);
    return String.format("0x%0" + (ValueLayout.ADDRESS.byteSize() * 2) + "X", value.address());
}
```

In JDK 20 release I get an error:
![image](https://user-images.githubusercontent.com/26604994/188499156-a1c19a0d-cdea-430d-8866-7f99e0fe7f8b.png)


Trying to do `(MemorySegment)` cast fails:

![image](https://user-images.githubusercontent.com/26604994/188499194-0e319c83-ee4d-4ff4-969f-f5ff2ca82587.png)

But if I make it take `Addressable` instead, then it works:
![image](https://user-images.githubusercontent.com/26604994/188499260-2e86662c-d317-43c8-a94d-ea8695de75e3.png)

Reply:
Note that the API in the Panama repo is different - MemoryAddress (and Addressable) have been removed as part of:
https://github.com/openjdk/panama-foreign/pull/694

Reply:
Oh, well that'd make sense why it didn't work OOTB then, thanks for the info!